// @vue/eslint-config-typescript/dist/index.cjs
'use strict';

var tseslint = require('typescript-eslint');
var tseslintParser = require('@typescript-eslint/parser');
var pluginVue = require('eslint-plugin-vue');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var tseslint__namespace = /*#__PURE__*/_interopNamespaceDefault(tseslint);
var tseslintParser__namespace = /*#__PURE__*/_interopNamespaceDefault(tseslintParser);

function createConfig({
  extends: configNamesToExtend = ["recommended"],
  supportedScriptLangs = { ts: true, tsx: false, js: false, jsx: false }
} = {}) {
  const mayHaveJsxInSfc = supportedScriptLangs.jsx || supportedScriptLangs.tsx;
  const needsTypeAwareLinting = configNamesToExtend.some(
    (name) => name.endsWith("-type-checked")
  );
  if (needsTypeAwareLinting && mayHaveJsxInSfc) {
    throw new Error(
      "Type-aware linting is not supported in Vue SFCs with JSX syntax. Please disable type-aware linting or set `supportedScriptLangs.jsx` and `supportedScriptLangs.tsx` to `false`."
    );
  }
  const noProjectServiceForVue = mayHaveJsxInSfc;
  const projectServiceConfigs = [];
  if (noProjectServiceForVue) {
    projectServiceConfigs.push({
      name: "vue-typescript/project-service-for-vue",
      files: ["*.vue", "**/*.vue"],
      languageOptions: {
        parserOptions: {
          projectService: false
        }
      }
    });
  }
  return tseslint__namespace.config(
    ...configNamesToExtend.map((configName) => tseslint__namespace.configs[configName]).flat().map(
      (config) => config.files && config.files.includes("**/*.ts") ? {
        ...config,
        files: [...config.files, "**/*.vue"]
      } : config
    ),
    ...pluginVue.configs["flat/base"],
    {
      name: "vue-typescript/setup",
      files: ["*.vue", "**/*.vue"],
      languageOptions: {
        parserOptions: {
          parser: {
            // Fallback to espree for js/jsx scripts, as well as SFCs without scripts
            // for better performance.
            js: "espree",
            jsx: "espree",
            ts: tseslintParser__namespace,
            tsx: tseslintParser__namespace
            // Leave the template parser unspecified,
            // so that it could be determined by `<script lang="...">`
          },
          ecmaFeatures: {
            jsx: mayHaveJsxInSfc
          },
          extraFileExtensions: ["vue"]
        }
      },
      rules: {
        "vue/block-lang": [
          "error",
          {
            script: {
              lang: Object.keys(supportedScriptLangs).filter(
                (lang) => supportedScriptLangs[lang]
              ),
              allowNoLang: supportedScriptLangs.js
            }
          }
        ]
      }
    },
    ...projectServiceConfigs
  );
}

module.exports = createConfig;
